lib/rubocop/cop/rspec/subject_declaration.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Ensure that subject is defined using subject helper.
+      #
+      # @example
+      #
+      #   # bad
+      #   let(:subject) { foo }
+      #   let!(:subject) { foo }
+      #   subject(:subject) { foo }
+      #   subject!(:subject) { foo }
+      #
+      #   # bad
+      #   block = -> {}
+      #   let(:subject, &block)
+      #
+      #   # good
+      #   subject(:test_subject) { foo }
+      #
+      class SubjectDeclaration < Base
+        MSG_LET = 'Use subject explicitly rather than using let'
+        MSG_REDUNDANT = 'Ambiguous declaration of subject'
+
+        # @!method let_subject?(node)
+        def_node_matcher :let_subject?, <<~PATTERN
+          (send nil? {#Helpers.all} {(sym :subject) (str "subject")} ...)
+        PATTERN
+
+        # @!method redundant_subject?(node)
+        def_node_matcher :redundant_subject?, <<~PATTERN
+          (send nil? {#Subjects.all} {(sym :subject) (str "subject")} ...)
+        PATTERN
+
+        def on_send(node)
+          let_subject?(node) do
```

Comment:
Alternatively, you can combine the matchers, and just check the captured node method name to determine the message